File: android/app/src/main/java/com/buddhist/uposatha/audio/AudioPlayerService.java
package com.buddhist.uposatha.audio;

import android.app.PendingIntent;
import android.content.Intent;
import androidx.annotation.Nullable;
import androidx.media3.common.AudioAttributes;
import androidx.media3.common.C;
import androidx.media3.common.MediaItem;
import androidx.media3.common.Player;
import androidx.media3.exoplayer.ExoPlayer;
import androidx.media3.session.MediaSession;
import androidx.media3.session.MediaSessionService;
import android.util.Log;

public class AudioPlayerService extends MediaSessionService {
    private static final String TAG = "AudioPlayerService";
    private ExoPlayer player;
    private MediaSession mediaSession;

    @Override
    public void onCreate() {
        super.onCreate();
        Log.d(TAG, "onCreate: Initializing ExoPlayer and MediaSession");

        // Initialize ExoPlayer
        player = new ExoPlayer.Builder(this)
                .setAudioAttributes(
                        new AudioAttributes.Builder()
                                .setContentType(C.AUDIO_CONTENT_TYPE_MUSIC)
                                .setUsage(C.USAGE_MEDIA)
                                .build(),
                        true
                )
                .setHandleAudioBecomingNoisy(true)
                .build();

        // Initialize MediaSession
        // In a real app, you might want a PendingIntent to open the Activity
        mediaSession = new MediaSession.Builder(this, player).build();
        
        player.addListener(new Player.Listener() {
            @Override
            public void onPlaybackStateChanged(int playbackState) {
                Log.d(TAG, "onPlaybackStateChanged: " + playbackState);
                if (playbackState == Player.STATE_ENDED) {
                    // Handle end of queue if needed
                }
            }

            @Override
            public void onPlayerError(androidx.media3.common.PlaybackException error) {
                Log.e(TAG, "onPlayerError: " + error.getMessage());
            }
        });
    }

    @Nullable
    @Override
    public MediaSession onGetSession(MediaSession.ControllerInfo controllerInfo) {
        return mediaSession;
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        return super.onStartCommand(intent, flags, startId);
    }

    @Override
    public void onDestroy() {
        Log.d(TAG, "onDestroy: Releasing resources");
        if (player != null) {
            player.release();
            player = null;
        }
        if (mediaSession != null) {
            mediaSession.release();
            mediaSession = null;
        }
        super.onDestroy();
    }
}

File: android/app/src/main/kotlin/com/buddhist/uposatha/audio/AudioPlayerService.kt
package com.buddhist.uposatha.audio

import android.app.PendingIntent
import android.content.Intent
import android.os.Build
import android.util.Log
import androidx.core.app.NotificationCompat
import androidx.media3.common.AudioAttributes
import androidx.media3.common.C
import androidx.media3.common.PlaybackException
import androidx.media3.common.Player
import androidx.media3.exoplayer.ExoPlayer
import androidx.media3.session.MediaSession
import androidx.media3.session.MediaSessionService
import androidx.media3.common.util.UnstableApi
import android.media.audiofx.LoudnessEnhancer
import com.buddhist.uposatha.MainActivity

class AudioPlayerService : MediaSessionService() {
    private var player: ExoPlayer? = null
    private var mediaSession: MediaSession? = null
    private var loudnessEnhancer: LoudnessEnhancer? = null

    companion object {
        private const val TAG = "AudioPlayerService"
        private const val CHANNEL_ID = "uposatha_audio_channel"
        private const val NOTIFICATION_ID = 1001
    }

    override fun onCreate() {
        super.onCreate()
        Log.d(TAG, "onCreate: Initializing ExoPlayer and MediaSession")

        player = ExoPlayer.Builder(this)
            .setAudioAttributes(
                AudioAttributes.Builder()
                    .setContentType(C.AUDIO_CONTENT_TYPE_MUSIC)
                    .setUsage(C.USAGE_MEDIA)
                    .build(),
                true
            )
            .setHandleAudioBecomingNoisy(true)
            .build()

        val intent = Intent(this, MainActivity::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
        }
        val pendingIntent = PendingIntent.getActivity(
            this, 0, intent, PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT
        )

        mediaSession = MediaSession.Builder(this, player!!)
            .setSessionActivity(pendingIntent)
            .build()

        player?.addListener(object : Player.Listener {
            @androidx.annotation.OptIn(UnstableApi::class)
            override fun onPlaybackStateChanged(playbackState: Int) {
                Log.d(TAG, "onPlaybackStateChanged: $playbackState")
                if (playbackState == Player.STATE_READY && player?.playWhenReady == true) {
                    startForegroundWithNotification()
                    if (loudnessEnhancer == null) {
                        try {
                            val audioSessionId = player?.audioSessionId ?: C.AUDIO_SESSION_ID_UNSET
                            if (audioSessionId != C.AUDIO_SESSION_ID_UNSET) {
                                loudnessEnhancer = LoudnessEnhancer(audioSessionId).apply {
                                    setTargetGain(1000) // 10dB boost
                                    enabled = true
                                }
                                Log.d(TAG, "LoudnessEnhancer enabled for session: $audioSessionId")
                            }
                        } catch (e: Exception) {
                            Log.e(TAG, "Failed to initialize LoudnessEnhancer", e)
                        }
                    }
                } else if (playbackState == Player.STATE_IDLE || playbackState == Player.STATE_ENDED) {
                    stopForeground(STOP_FOREGROUND_DETACH)
                }
            }

            override fun onPlayerError(error: PlaybackException) {
                Log.e(TAG, "onPlayerError: code=${PlaybackException.getErrorCodeName(error.errorCode)} (${error.errorCode})", error)
            }
        })
    }

    private fun startForegroundWithNotification() {
        val builder = NotificationCompat.Builder(this, CHANNEL_ID)
            .setSmallIcon(android.R.drawable.ic_media_play)
            .setContentTitle("Dhamma Talk")
            .setContentText("Playing audio...")
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .setOngoing(true)

        startForeground(NOTIFICATION_ID, builder.build())
    }

    override fun onGetSession(controllerInfo: MediaSession.ControllerInfo): MediaSession? = mediaSession

    override fun onDestroy() {
        Log.d(TAG, "onDestroy: Releasing resources")
        player?.release()
        player = null
        mediaSession?.release()
        mediaSession = null
        loudnessEnhancer?.release()
        loudnessEnhancer = null
        super.onDestroy()
    }
}
